### DISCUSSION TITLE
journal paper on perf analysis of cpython

### DISCUSSION THREAD

**Discussion by cfbolz** [faster-cpython/ideas#213] [General]
This paper appeared recently, and I thought people here might be interested: https://www.sciencedirect.com/science/article/abs/pii/S0167642321001520

It very carefully instruments CPython with perf to find out where time is spent. It's really a very interesting read, I enjoyed it a lot.

The corresponding benchmarking and instrumentation infrastructure is here: https://github.com/zq1997/cpython-overhead-evaluation (👋👋 @zq1997)

Unfortunately the paper is paywalled, if you can't find a way to get access, feel free to send me a private mail.

**Comment by ericsnowcurrently** [comment]
That does sound interesting.  The abstract certainly is promising.  Do you think it would be worth getting on a call with one or more of the authors?

**Comment by gvanrossum** [comment]
@cfboltz: I sent you an email, but I don't know of you monitor that inbox any more (I had to search hard to find an email address for you).

**Reply by cfbolz** [reply]
I don't know what happened, I have gotten your email and have replied right away. Both Eric and Mark should have copies as well though!

**Comment by gvanrossum** [comment]
NM, I got a copy from a coworker. It's interesting indeed! Every core dev should read it (and many have their work cited, even Brett's PhD thesis).

**Comment by gvanrossum** [comment]
An interesting idea from the paper (after observing that the python_startup* benchmarks waste a lot of time in GC):

> Perhaps we can set a higher GC threshold for better startup performance and restore it to the normal value after startup.
> It can improve the startup performance by about 3.0% without extra memory cost. The improvement is small, but it has
> almost no side effects and is easy to implement.

**Comment by gvanrossum** [comment]
They also seem to have formed the opinion that the threaded code (i.e., computed gotos to avoid the switch) is not making as much of a difference as was thought (or maybe modern processors are better at branch prediction).

Lots of other things to conclude, I'll try to stop spamming this issue.
